networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -0,0 +1,716 @@
+
+    Time complexity:
+    It has a running time $O(|E||V|^2 + m * n * |V|)$ when a greedy
+    algorithm is used to construct an initial solution, otherwise
+    it has a running time $O(m * n * |V|)$ where $m$ and $n$ are the number
+    of iterations of outer and inner loop respectively.
+
+    For more information and how algorithm is inspired see:
+    http://en.wikipedia.org/wiki/Threshold_accepting
+
+    See Also
+    --------
+    simulated_annealing_tsp
+
+    """
+    if cycle is None:
+        # Construct an initial solution using a greedy algorithm.
+        cycle = greedy_tsp(G, weight=weight, source=source)
+        if G.number_of_nodes() == 2:
+            return cycle
+
+        cost = sum(G[u][v][weight] for u, v in zip(cycle, cycle[1:]))
+    else:
+        # Find the cost of initial solution and make the essential checks for graph.
+        if any(v not in G[u] for u in G for v in G if u != v):
+            raise nx.NetworkXError(
+                "Not a complete graph. All node pairs must have edges."
+            )
+        if not nx.is_weighted(G, weight=weight):
+            raise nx.NetworkXError("Given graph is not weighted.")
+
+        if G.number_of_nodes() == 2:
+            neighbor = list(G.neighbors(source))[0]
+            return [source, neighbor, source]
+
+        cost = sum(G[u][v][weight] for u, v in zip(cycle, cycle[1:]))
+
+    count = 0
+    best_cycle = cycle[:]
+    best_cost = cost
+    while count <= tolerance:
+        count += 1
+        accepted = False
+        for i in range(iterations):
+            adj_sol = _apply_move(cycle, move, seed)
+            adj_cost = sum(G[u][v][weight] for u, v in zip(cycle, cycle[1:]))
+            delta = adj_cost - cost
+            if delta <= threshold:
+                accepted = True
+
+                # Set current solution the adjacent solution.
+                cycle = adj_sol
+                cost = adj_cost
+
+                if cost < best_cost:
+                    count = 0
+                    best_cycle = cycle[:]
+                    best_cost = cost
+        if accepted:
+            threshold -= threshold * alpha
+
+    return best_cycle
+
+
+def _apply_move(soln, move, seed):
+    """
+    Apply a move to a solution to generate a neighbor solution.
+
+    Parameters
+    ----------
+    soln : list of nodes
+        Current solution (list of nodes)
+
+    move : string
+        Move to be applied
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    The solution after move is applied. (A neighbor solution.)
+    """
+    a = seed.randint(1, len(soln) - 2)
+    listb = list(range(1, a)) + list(range(a + 1, len(soln) - 1))
+    b = seed.choice(listb)
+    if move == "1-1":
+        soln[a], soln[b] = soln[b], soln[a]
+    elif move == "1-0":
+        soln.insert(b, soln.pop(a))
```

Comment:
AFAICT there is no check for an invalid move - either add one here:

```suggestion
        soln.insert(b, soln.pop(a))
    else:
        raise ValueError(
            f"Invalid move: {move}. Expected one of ('1-1', '1-0')"
        )
```

Or add a check in the top-level SA and TA functions.

Reply:
Yes!  This is important because a mis-spelling would proceed without ever making a move!
It's probably good to put in SA and TA functions. But it's cheap to put it here even if it is used in an inner loop. The `if` already exists... it's just the else that's added.